code/go/0chain.net/blobbercore/automigration/automigration.go

```diff
@@ -0,0 +1,159 @@
+	new(allocation.WritePool),
+	new(readmarker.ReadMarkerEntity),
+	new(writemarker.WriteMarkerEntity),
+	new(writemarker.WriteLock),
+	new(stats.FileStats),
+}
+
+func AutoMigrate(pgDB *gorm.DB) error {
+	if err := createUser(pgDB); err != nil {
+		return err
+	}
+
+	if err := createDB(pgDB); err != nil {
+		return err
+	}
+
+	if err := grantPrivileges(pgDB); err != nil {
+		return err
+	}
+
+	d, err := pgDB.DB()
+	if err != nil {
+		return err
+	}
+
+	if err := d.Close(); err != nil {
+		return err
+	}
+
+	if err := datastore.GetStore().Open(); err != nil {
+		return err
+	}
+
+	db := datastore.GetStore().GetDB()
+
+	return migrateSchema(db)
+}
+
+func createDB(db *gorm.DB) error {
+	// check if db exists
+	dbstmt := fmt.Sprintf("SELECT datname, oid FROM pg_database WHERE datname = '%s';", config.Configuration.DBName)
+	rs := db.Raw(dbstmt)
+	if rs.Error != nil {
+		return rs.Error
+	}
+
+	var result struct {
+		Datname string
+	}
+
+	if rs.Scan(&result); len(result.Datname) == 0 {
+		stmt := fmt.Sprintf("CREATE DATABASE %s;", config.Configuration.DBName)
+		if rs := db.Exec(stmt); rs.Error != nil && rs.Error.Error() != fmt.Sprintf("pq: database \"%s\" already exists", config.Configuration.DBName) {
+			return rs.Error
+		}
+	}
+	return nil
+}
+
+func createUser(db *gorm.DB) error {
+	usrstmt := fmt.Sprintf("SELECT usename, usesysid FROM pg_catalog.pg_user WHERE usename = '%s';", config.Configuration.DBUserName)
+	rs := db.Raw(usrstmt)
+	if rs.Error != nil {
+		return rs.Error
+	}
+
+	var result struct {
+		Usename string
+	}
+
+	if rs.Scan(&result); len(result.Usename) == 0 {
+		stmt := fmt.Sprintf("CREATE USER %s WITH ENCRYPTED PASSWORD '%s';", config.Configuration.DBUserName, config.Configuration.DBPassword)
+		if rs := db.Exec(stmt); rs.Error != nil && rs.Error.Error() != fmt.Sprintf("pq: role \"%s\" already exists", config.Configuration.DBUserName) {
+			return rs.Error
+		}
+	}
+	return nil
+}
+
+func grantPrivileges(db *gorm.DB) error {
+	stmts := []string{
+		fmt.Sprintf("GRANT ALL PRIVILEGES ON DATABASE %s TO %s;", config.Configuration.DBName, config.Configuration.DBUserName),
+		fmt.Sprintf("GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO %s;", config.Configuration.DBUserName),
+		fmt.Sprintf("GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO %s;", config.Configuration.DBUserName),
+	}
+	for _, stmt := range stmts {
+		rs := db.Raw(stmt)
+		if rs.Error != nil {
+			return rs.Error
+		}
+	}
+	return nil
+}
+
+func migrateSchema(db *gorm.DB) error {
+	if config.Configuration.DBDropAllTables {
+		tablesToKeep := make(map[string]struct{})
+		for _, tableName := range config.Configuration.DBTablesToKeep {
+			tablesToKeep[tableName] = struct{}{}
+		}
+
+		for _, tbMdl := range tableModels {
+			if _, ok := tablesToKeep[tbMdl.TableName()]; ok {
+				continue
+			}
+			if err := db.Migrator().DropTable(tbMdl); err != nil {
+				return err
+			}
+		}
+	} else {
+		for _, tableName := range config.Configuration.DBTablesToDrop {
```

Comment:
If tables exist in DBTablesToDrop and DBTablesToKeep, should they be dropped or kept?

Reply:
It is based on `DBDropAllTables` value. If it is `true` it will take values from `DBTablesToKeep` otherwise from `DBTablesToDrop`.

Reply:
We'd better have code to check if there are tables overlapped, or define the priority about how to process it if overlapped. For example, suppose tables overlap is allowed, and  `DBTableToKeep` has higher priority, then we should exclude them from `DBTablesToDrop`, and drop the remaining ones only.

Reply:
So this means we don't need to use `DBDropAllTables` field.

Reply:
Yes, seems like the `DBDropAllTables` is not necessary. As the `DBTablesToKeep` and `DBTablesToDrop` can present the drop all case. 

One issue about **not** using `DBDropAllTables` is if we have a lot of tables, it's annoying to put all tables to the `DBTablesToDrop`, then the `DBDropAllTables` would be helpful, but that will need us to have the way to get all the tables, and then drop all of them.